sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveUtils.scala

```diff
@@ -397,27 +399,86 @@ private[spark] object HiveUtils extends Logging {
         barrierPrefixes = hiveMetastoreBarrierPrefixes,
         sharedPrefixes = hiveMetastoreSharedPrefixes)
     } else {
-      // Convert to files and expand any directories.
-      val jars =
-        hiveMetastoreJars
-          .split(File.pathSeparator)
-          .flatMap {
-          case path if new File(path).getName == "*" =>
-            val files = new File(path).getParentFile.listFiles()
-            if (files == null) {
-              logWarning(s"Hive jar path '$path' does not exist.")
+
+      def addLocalHiveJars(file: File): Seq[URL] = {
+        if (file.getName == "*") {
+          val files = file.getParentFile.listFiles()
+          if (files == null) {
+            logWarning(s"Hive jar path '${file.getPath}' does not exist.")
+            Nil
+          } else {
+            files.filter(_.getName.toLowerCase(Locale.ROOT).endsWith(".jar")).map(_.toURL).toSeq
+          }
+        } else {
+          file.toURL :: Nil
+        }
+      }
+
+      def checkRemoteHiveJars(path: String): Seq[URL] = {
+        try {
+          val hadoopPath = new Path(path)
+          val fs = hadoopPath.getFileSystem(hadoopConf)
+          if (hadoopPath.getName == "*") {
+            val parent = hadoopPath.getParent
+            if (!fs.exists(parent)) {
+              logWarning(s"Hive Jar ${path} does not exist.")
+              Nil
+            } else if (!fs.getFileStatus(parent).isDirectory) {
+              logWarning(s"Hive Jar ${parent} is not a directory.")
               Nil
             } else {
-              files.filter(_.getName.toLowerCase(Locale.ROOT).endsWith(".jar")).toSeq
+              fs.listStatus(parent).map(_.getPath.toUri.toURL)
             }
-          case path =>
-            new File(path) :: Nil
+          } else {
+            if (!fs.exists(hadoopPath)) {
+              logWarning(s"Hive Jar ${path} does not exist.")
+              Nil
+            } else if (fs.getFileStatus(hadoopPath).isDirectory) {
+              logWarning(s"Hive Jar ${path} not allow directory without `*`")
+              Nil
+            } else {
+              // Since tar/tar.gz file we can't know it's final path yet, not support it
+              hadoopPath.toUri.toURL :: Nil
+            }
+          }
+        } catch {
+          case NonFatal(e) =>
+            logError(s"Failed to find $path to Hive Jars", e)
+            Nil
         }
-          .map(_.toURI.toURL)
+      }
+
+      // Convert to files and expand any directories.
+      val jars =
+        hiveMetastoreJars
+          .split(";")
```

Comment:
Why do we change this from `File.pathSeparator` to `;`?

Reply:
> Why do we change this from `File.pathSeparator` to `;`?

Since in unix system, `File.pathSeparator` is `:` will split hdfs path `hdfs://xxx/xxx`

Reply:
File separator is OS dependent.. lets don't change the behaviour

Reply:
> Path separator is OS dependent.. lets don't change the behaviour

But a lot place we use comma separator to split file path, right? Also I am confused that why we must need to use this Path separator here since what we config is jar file list or `*`.

or you mean use will pass config like `hadoop classpath`, but seem no `hive classpath`

Reply:
ping @dongjoon-hyun @wangyum @cloud-fan WDYT about this problem

Reply:
Hey, yes using comma might be a good way but now you're breaking the user applications that use `:`. Let's don't break.
I think path separator is taken after the regular classpath behaviour when you run jars like `java -cp`.